src/rotation.rs

```diff
@@ -0,0 +1,687 @@
+        let m32 = jk - ri;
+        let m33 = one - (ii + jj);
+
+        TypedTransform3D::row_major(
+            m11,  m12,  m13, zero,
+            m21,  m22,  m23, zero,
+            m31,  m32,  m33, zero,
+            zero, zero, zero, one,
+        )
+    }
+
+    /// Returns a rotation representing the other rotation followed by this rotation.
+    pub fn pre_rotate<NewSrc>(&self, other: &TypedRotation3D<T, NewSrc, Src>) -> TypedRotation3D<T, NewSrc, Dst> {
+        debug_assert!(self.is_normalized());
+        TypedRotation3D::quaternion(
+            self.i * other.r + self.r * other.i + self.j * other.k - self.k * other.j,
+            self.j * other.r + self.r * other.j + self.k * other.i - self.i * other.k,
+            self.k * other.r + self.r * other.k + self.i * other.j - self.j * other.i,
+            self.r * other.r - self.i * other.i - self.j * other.j - self.k * other.k,
+        )
+    }
+
+    /// Returns a rotation representing this rotation followed by the other rotation.
+    #[inline]
+    pub fn post_rotate<NewDst>(&self, other: &TypedRotation3D<T, Dst, NewDst>) -> TypedRotation3D<T, Src, NewDst> {
+        other.pre_rotate(self)
+    }
+
+    /// Returns the vector part (i, j, k) of this quaternion.
+    #[inline]
+    pub fn vector_part(&self) -> Vector3D<T> { vec3(self.i, self.j, self.k) }
+
+    // add, sub and mul are used internally for intermediate computation but aren't public
+    // because they don't carry real semantic meanings (I think?).
+
+    #[inline]
+    fn add(&self, other: Self) -> Self {
+        Self::quaternion(
+            self.i + other.i,
+            self.j + other.j,
+            self.k + other.k,
+            self.r + other.r,
+        )
+    }
+
+    #[inline]
+    fn sub(&self, other: Self) -> Self {
+        Self::quaternion(
+            self.i - other.i,
+            self.j - other.j,
+            self.k - other.k,
+            self.r - other.r,
+        )
+    }
+
+    #[inline]
+    fn mul(&self, factor: T) -> Self {
+        Self::quaternion(
+            self.i * factor,
+            self.j * factor,
+            self.k * factor,
+            self.r * factor,
+        )
+    }
+}
+
+impl<T: fmt::Debug, Src, Dst> fmt::Debug for TypedRotation3D<T, Src, Dst> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Quat({:?}*i + {:?}*j + {:?}*k + {:?})", self.i, self.j, self.k, self.r)
+    }
+}
+
+impl<T: fmt::Display, Src, Dst> fmt::Display for TypedRotation3D<T, Src, Dst> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Quat({}*i + {}*j + {}*k + {})", self.i, self.j, self.k, self.r)
+    }
+}
+
+impl<T, Src, Dst> ApproxEq<T> for TypedRotation3D<T, Src, Dst>
+where
+    T: Copy + Neg<Output=T> + ApproxEq<T>
+{
+    fn approx_epsilon() -> T {
+        T::approx_epsilon()
+    }
+
+    fn approx_eq(&self, other: &Self) -> bool {
+        (
+            self.i.approx_eq(&other.i)
+            && self.j.approx_eq(&other.j)
```

Comment:
nit: could be shorter with using `get_vector`

Reply:
D'oh! It would if approx_eq was sensibly implemented for the vector type, but the latter implements ApproxEq<Self> instead of ApproxEq<T> as it should.

I'll fix that later but that's techinally a breaking change so I'll do it whenever we have another breaking change to justify it.